### ISSUE TITLE
e2e flake: Namespace wasn't deleted correctly

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#22045]
It seems there is some bug around namespace controller. What happened is:

    08:05:39 STEP: Collecting events from namespace "e2e-tests-density-b5lqr".
    08:05:39 STEP: Destroying namespace "e2e-tests-density-b5lqr" for this suite.
    09:05:40 Feb 25 09:05:40.322: FAIL: Couldn't delete ns "e2e-tests-density-b5lqr": namespace e2e-tests-density-b5lqr was not deleted within limit: timed out waiting for the condition, pods remaining: []

So basically, namespace wasn't deleted for the whole hour.
Then the next test actually timed out

    09:05:40 Feb 25 09:05:40.997: INFO: >>> testContext.KubeConfig: /jenkins-master-data/jobs/kubernetes-kubemark-gce/workspace/kubernetes/test/kubemark/kubeconfig.loc
    09:05:40 
    09:05:41 Feb 25 09:05:41.182: INFO: Waiting for terminating namespaces to be deleted...
    09:46:36 
    Build timed out (after 120 minutes). Marking the build as failed.

So in total, namespace wasn't deleted for more than 90 minutes (note that usually it takes ~2-3 minutes for density test).

https://storage.googleapis.com/kubernetes-jenkins/logs/kubernetes-kubemark-gce/5970/build-log.txt

The problem is that we can't really debug it, since we don't have logs from k8s components if build times out. - @ixdy @spxtr @kubernetes/goog-testing

**Comment by davidopp** [comment]
@derekwaynecarr

**Comment by wojtek-t** [comment]
Just one detail - my parallelization of namespace deletion is not yet merged - so it's not because of that.

**Comment by ncdc** [comment]
@ixdy I noticed that the artifacts directory in the storage bucket is missing all the apiserver/kubelet/etc logs. Is it possible to modify the test infra to upload those in a situation like this?

**Comment by justinsb** [comment]
Also, there's this comment in `test/e2e/util.go` around namespace deletion, which suggests that the problem could be anywhere - the namespace controller could be perfectly correct, just waiting on something else:


```
        // TODO: Since we don't have support for bulk resource deletion in the API,
        // while deleting a namespace we are deleting all objects from that namespace
        // one by one (one deletion == one API call). This basically exposes us to
        // throttling - currently controller-manager has a limit of max 20 QPS.
        // Once #10217 is implemented and used in namespace-controller, deleting all
        // object from a given namespace should be much faster and we will be able
        // to lower this timeout.
        // However, now Density test is producing ~26000 events and Load capacity test
        // is producing ~35000 events, thus assuming there are no other requests it will
        // take ~30 minutes to fully delete the namespace. Thus I'm setting it to 60
        // minutes to avoid any timeouts here.
        timeout := 60 * time.Minute
```

**Comment by ncdc** [comment]
That comment is a bit out of date. We do have delete-collection in the apiserver now. Its implementation isn't optimal (it's 1 request to etcd per object to delete), but the difference is it's a single call to the apiserver instead of multiple. From what I can see, we use that everywhere when deleting a namespace except for services (delete collection not yet implemented for them) and pods (need to check grace period per pod).

There is also Derek's open PR to convert the namespace deletion code to use the generic discovery client. Not sure how that may/may not impact timing. #21400

**Comment by wojtek-t** [comment]
Yes - I agree. This has been fixed (currently it takes ~2-3 minutes to delete namespace in 100-node cluster, where this failure come from).

@ncdc - regarding artifacts - yes we definitely want this. There is a long standing P0 bug for it:
https://github.com/kubernetes/kubernetes/issues/21722
[it's more for failing kube-up, timing out builds are also part of it]

**Comment by derekwaynecarr** [comment]
It's possible a side-effect of this:

https://github.com/kubernetes/kubernetes/issues/21406

I will take a look to see if I can improve the error handling on a sync call and/or log more detail to know what was the issue.

**Comment by wojtek-t** [comment]
@derekwaynecarr - yeah that's definitely possible; previously in the tests it happened that there was some connection problem when connecting to etcd - it might have repeated.

**Comment by bgrant0607** [comment]
I don't know if this is related, but I just noticed this:
```
[It] should scale a job down
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/job.go:161
STEP: Creating a job
STEP: Ensuring active pods == startParallelism
STEP: scale job down
STEP: Ensuring active pods == endParallelism
[AfterEach] Job
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:84
Feb 26 15:52:34.686: INFO: Waiting up to 1m0s for all nodes to be ready
STEP: Destroying namespace "e2e-tests-job-88kfc" for this suite.
Feb 26 15:52:39.707: INFO: Pod e2e-tests-job-88kfc scale-down-wmgrj on node  remains, has deletion timestamp <nil>
Feb 26 15:52:39.707: FAIL: Couldn't delete ns "e2e-tests-job-88kfc": pods remained within namespace e2e-tests-job-88kfc after deletion: [scale-down-wmgrj]


â€¢ Failure in Spec Teardown (AfterEach) [79.637 seconds]
Job
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/job.go:198
  should scale a job down [AfterEach]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/job.go:161

  Feb 26 15:52:39.707: Couldn't delete ns "e2e-tests-job-88kfc": pods remained within namespace e2e-tests-job-88kfc after deletion: [scale-down-wmgrj]

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:156
```
kubernetes-jenkins/pr-logs/pull/22078/kubernetes-pull-build-test-e2e-gce/30722

**Comment by derekwaynecarr** [comment]
This PR should merge soon that updates namespace deletion and adds additional validation that a LIST returns an empty set for each API resource.  

https://github.com/kubernetes/kubernetes/pull/21400

I want to keep this open until we verify we no longer see the flake, and have been looking at where else we can add logging in the interim to determine where the root cause problem is since it's not clear to me it's in namespace controller versus somewhere else.

**Comment by derekwaynecarr** [comment]
my other thought if this continues to happen is the admission controller in NamespaceLifecycle is not getting notification of the namespace phase change and deletion fast enough , and a pod is being admitted even after a namespace no longer exists by another latent controller.  If that is the source I am not sure of the resolution.   We could require an admission controller to take part in namespace finalization but that would slow down namespace deletion and also seems wrong if we scale the apiserver horizontally.

**Comment by derekwaynecarr** [comment]
@wojtek-t - have you seen any more occurrences of this flake? 

my other theory is that a pod was not being terminated due to the root cause of https://github.com/kubernetes/kubernetes/issues/22025

if the flake is no longer occurring, any more objections to closing?

**Comment by wojtek-t** [comment]
No - I didn't see any other occurrences of it. Let's close it - we can reopen if it happens again.

**Comment by piosz** [comment]
It happened again:
```
Services should be able to change the type and ports of a service

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:101 Mar 8 04:00:45.191: Couldn't delete ns "e2e-tests-services-pqfc6": pods remained within namespace e2e-tests-services-pqfc6 after deletion: [mutability-test-e5am2]
```
https://storage.googleapis.com/kubernetes-jenkins/logs/kubernetes-e2e-gke/2440/build-log.txt
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gke/2440/

**Comment by wojtek-t** [comment]
Unfortunately we don't have logs in GKE...

**Comment by wojtek-t** [comment]
I mean logs from apiserver

**Comment by derekwaynecarr** [comment]
I suspect this has a latent cache when this flake occurs that is letting in a resource spawned by a controller after the namespace is deleted:

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/namespace/lifecycle/admission.go#L93

I am not sure what to do here... suggestions welcome.  

* changing the namespace controller to do something order specific should not be needed
* in theory, we could add a delay to let assumed caches propagate so admission is blocking new creates, but that also is less than ideal.

**Comment by wojtek-t** [comment]
@derekwaynecarr - I agree that this is possible. But when I looked into the test I didn't see any reasonable explanation what object can be created/updated in the namespace after the request to delete it. Any thoughts on it?

**Comment by derekwaynecarr** [comment]
@wojtek-t - I think we can reduce the flake by having the ns controller create an Event in a terminating namespace.  Once it gets a response back from the server that says the Event cannot be created, it knows at least one admission controller (assuming a clustered api server) has had its local cache propagated, and no further creates would be let in to that namespace.  Ideally, we would have a way to check each endpoint that exposes the api server to try individually if we wanted a bullet proof solution.

**Comment by derekwaynecarr** [comment]
@wojtek-t - the test in question creates multiple replication controllers:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/service.go#L452

we could modify the test to manually delete the replication controller as well.

**Comment by derekwaynecarr** [comment]
or discover scaled resources and scale those to 0 and block until the status reports 0 for each scaled resource.  not sure how easy that is to do generically.

**Comment by derekwaynecarr** [comment]
This will be my top priority on Monday pending more thoughts on how to do this best.

**Comment by bgrant0607** [comment]
Will comment later.

**Comment by thockin** [comment]
I thought we marked the namespace as Terminating, which would fail any subsequent POST operations, which should free us up to delete in any order (since we have no referential integrity or delete inhibition)?  What did I get wrong?

**Comment by derekwaynecarr** [comment]
@thockin - we do.  the issue best that i can tell is that when this flake occurs the cache in that admission controller is stale.  so while namespace controller has observed it should delete, the admission controller that is blocking creates has not yet made that observation.

**Comment by wojtek-t** [comment]
>
@wojtek-t - I think we can reduce the flake by having the ns controller create an Event in a terminating namespace. Once it gets a response back from the server that says the Event cannot be created, it knows at least one admission controller (assuming a clustered api server) has had its local cache propagated, and no further creates would be let in to that namespace (for at least one server in the cluster). Ideally, we would have a way to check each endpoint that exposes the api server to try individually if we wanted a bullet proof solution, but that list of available endpoints is not readily passed into controllers today...

@derekwaynecarr - I think adding additional event is over-complication. We already have a cache with namespaces - we should just wait until "terminating" - we should just wait until state "terminating" for a given namespace is observed in this cache.
This still doesn't solve the problem with multiple apiserver, but it can be generalized exactly as you wrote above.
WDYT?

**Comment by derekwaynecarr** [comment]
@wojtek-t - there are 2 caches in separate processes at play here.  the namespace controller correctly sees that a namespace is terminating in its cache.  it then deletes all the things.  what it can't see is if the cache in admission control in the kube-apiserver process has seen the terminating state.  the event creator was a cheap way of knowing that it had seen the state transition in the admission cache since it would be told no by admission when the admission cache is updated.  the other alternative is to have the admission controller annotate a namespace and have the controller block on state change + annotation, but i dislike that. 

@bgrant0607 - did you have a chance to think about this anymore?

**Comment by derekwaynecarr** [comment]
Using the `Event` pattern to know when caches were updated, you cant really create an `Event` that is about a `Namespace` today because of a validation error, so if I go that route, we need to loosen up:

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/events.go#L36

**Comment by wojtek-t** [comment]
@derekwaynecarr - ok got it, thanks for explanation

>
 the other alternative is to have the admission controller annotate a namespace and have the controller block on state change + annotation, but i dislike that.

Why you don't like it? Actually it sounds pretty good (I would choose this one).
The solution with events also sound fine.

**Comment by derekwaynecarr** [comment]
@wojtek-t - i need to reason through what happens in the OpenShift case where we have two things interacting in namespace deletion.  it seems like you would need to set an annotation per namespace finalizer token that says admission has blocked new content creation.  thinking through that more, that works better than events since the event pattern only worked for core kube resources.

**Comment by wojtek-t** [comment]
I agree, that doesn't solve the problems, but it's not worse than events solution in my opinion (we have exaclty the same set of problems there).

**Comment by derekwaynecarr** [comment]
yeah, i think i have reasoned my way out of the event pattern since it doesn't help us in OpenShift where we have our own resources tied to namespace lifecycle.  the annotation pattern bugs me mostly because its an annotation and not a true field on the namespace, it makes putting policy around setting the annotation hard.

**Comment by bgrant0607** [comment]
Don't use events.

**Comment by derekwaynecarr** [comment]
Another option is that we have two finalizers on the namespace:

"kubernetes" - used by the controller
"kubernetes.admission" - used by admission

We set both on create on the namespace.  When the admission sees its cache update, we remove "kubernetes.admission" from finalizer list.  The controller would then know it can then do its delete cleanup, and then remove "kubernetes".

That puts some implied ordering between finalizer tokens, but gets the same benefit without conflating annotations.

**Comment by bgrant0607** [comment]
@derekwaynecarr I like 2 finalizers.

**Comment by bgrant0607** [comment]
In general, we definitely need an explicit ACK from admission control

**Comment by bgrant0607** [comment]
@derekwaynecarr I assume this isn't for 1.2?

**Comment by derekwaynecarr** [comment]
@bgrant0607 - I do not want to do a dual finalizer solution for 1.2.  Seems too risky this late in the process, and is still problematic when running multiple API servers behind a load balancer since there is no guarantee that because one admission controller ACK occured that the others had occurred.  So it's got problems.

I think for 1.2, we can reduce incidence of the flake by doing the following:

* Update `NamespaceLifecycle` to remove a `Namespace` from its local cache when it sees a `DELETE`
* Subsequent create/update operations would induce a live-lookup and force a refresh of the cache.
* That live lookup would capture the `Terminating` state.

I think that is fine for 1.2, thoughts?

I think longer term I prefer a pattern where a controller can discover the list of endpoints that satisfy the api-server, and it has a resource it can invoke that says 'could i do a create in this namespace' for each endpoint.

**Comment by derekwaynecarr** [comment]
I opened https://github.com/kubernetes/kubernetes/pull/22951 which should reduce incidence of the flake when running a single kube-apiserver since it will eliminate the watch notification delay by forcing a live-lookup of the latest namespace state after its deleted.

**Comment by wojtek-t** [comment]
This happened again:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/23373/kubernetes-pull-build-test-e2e-gce/33399/build-log.txt

**Comment by derekwaynecarr** [comment]
@wojtek-t - this is different.

Mar 23 07:17:33.181: FAIL: Couldn't delete ns "e2e-tests-kubectl-xh9eq": namespace e2e-tests-kubectl-xh9eq was not deleted within limit: timed out waiting for the condition, pods remaining: [frontend-9o032]

The old flake was that the namespace was deleted, but pods remained.

In this case, the namespace was not deleted because pods still remained.

**Comment by ncdc** [comment]
Can we open a new issue for this new failure condition?

**Comment by derekwaynecarr** [comment]
@ncdc I think for the above case, we need a bug that is focused more on the kubelet and why the pod was not terminated.  I suspect it could be similar to the same bug we are exploring here:

https://github.com/openshift/origin/issues/8176
